bazel/DEVELOPER.md

```diff
@@ -0,0 +1,225 @@
+
+## Adding files to the Envoy build
+
+All modules that make up the Envoy binary are statically linked at compile time.
+Many of the modules within Envoy have a pure virtual interface living in
+[`include/envoy`](../include/envoy), implementation sources in
+[`source`](../source), mocks in [`test/mocks`](../test/mocks) and
+unit/integration tests in [`test`](../test). The relevant `BUILD` files will
+require updating or to be added in these locations as you extend Envoy.
+
+As an example, consider adding the following interface in `include/envoy/foo/bar.h`:
+
+```c++
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/foo/baz.h"
+
+class Bar {
+public:
+  virtual ~Bar() {}
+
+  virtual void someThing() PURE;
+  ...
+```
+
+This would require the addition to `include/envoy/foo/BUILD` of the following target:
+
+```python
+envoy_cc_library(
+    name = "bar_interface",
+    hdrs = ["bar.h"],
+    deps = [
+        ":baz_interface",
+        "//include/envoy/buffer:buffer_interface",
+    ],
+)
+```
+
+This declares a new target `bar_interface`, where the convention is that pure
+virtual interfaces have their targets suffixed with `_interface`. The header
+`bar.h` is exported to other targets that depend on
+`//incude/envoy/foo:bar_interface`. The interface target itself depends on
+`baz_interface` (in the same directory, hence the relative Bazel label) and
+`buffer_interface`.
+
+In general, any header included via `#include` in a file belonging to the union
+of the `hdrs` and `srcs` lists for a Bazel target X should appear directly in
+the exported `hdrs` list for some target Y listed in the `deps` of X.
+
+Continuing the above example, the implementation of `Bar` might take place in
+`source/common/foo/bar_impl.h`, e.g.
+
+```c++
+#pragma once
+
+#include "envoy/foo/bar.h"
+
+class BarImpl : public Bar {
+...
+```
+
+and `source/common/foo/bar_impl.cc`:
+
+```c++
+#include "common/foo/bar_impl.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/foo/bar_internal.h"
+#include "common/foo/baz_impl.h"
+...
+```
+
+The corresponding target to be added to `source/common/foo/BUILD` would be:
+
+```python
+envoy_cc_library(
+    name = "bar_lib",
+    srcs = [
+        "bar_impl.cc",
+        "bar_internal.h",
+    ],
+    hdrs = ["bar_impl.h"],
+    deps = [
+        ":baz_lib",
+        "//include/envoy/foo:bar_interface",
+        "//source/common/buffer:buffer_lib",
+    ],
+)
+```
+
+By convention, Bazel targets for internal implementation libraries are suffixed
+with `_lib`.
+
+Similar to the above, a test mock target might be declared for `test/mocks/foo/mocks.h` in
+`test/mocks/foo/BUILD` with:
+
+```python
+envoy_cc_mock(
+    name = "foo_mocks",
+    srcs = ["mocks.cc"],
+    hdrs = ["mocks.h"],
+    deps = [
+        "//include/envoy/foo:bar_interface",
+        ...
+    ],
+)
+```
+
+Typically, mocks are provided for all interfaces in a directory in a single
+`mocks.{cc,h}` and corresponding `_mocks` Bazel target.
```

Comment:
Not sure if it is worth it to mention different cases like https://github.com/lyft/envoy/blob/master/test/mocks/upstream/BUILD, where there are mocks in files other than mocks.{cc,h}